### PR TITLE
Address #1185

### PR DESCRIPTION
**Pull request by NiklasGustafsson**
@shaltielshmid, @GilesBathgate -- please review.

### CODE REVIEW

File: test/TorchSharpTest/NN.cs [hunk 1]
```diff
@@ -4714,6 +4714,12 @@ public void TestLayerNorm()
                     Assert.Equal(ones.shape, pooled.shape);
                 }
             }
+            if (torch.cuda.is_available()) {
+                using var pool = LayerNorm(new long[] { 12, 28, 28 }, device: torch.CPU);
+
+                using var pool1 = pool.to(torch.CUDA);
+                Assert.Equal(DeviceType.CUDA, pool1.weight.device_type);
+            }
```
Comment (shaltielshmid) [inline test/TorchSharpTest/NN.cs:4722]: I don't think this unit test covers the bug reported. 
The issue was the LayerNorm not being recognized as a submodule, so perhaps placing it inside a sequential?

File: src/TorchSharp/NN/Module.cs [hunk 2]
```diff
@@ -251,7 +251,9 @@ private void _toEpilog(ScalarType? dtype, Device device)
                         // In addition, we need the new tensor to be a leaf to accumulate gradients, so if we didn't
                         // disable grad we would need to call .detach() on the moved tensor.
                         using (var d = torch.no_grad())
-                            p = new Parameter(param.to(paramType, device ?? param.device, disposeAfter: true), requiresGrad);
+                            p = new Parameter(
+                                param.to(paramType, device ?? param.device, disposeAfter: true).DetachFromDisposeScope(), requiresGrad)
+                                .DetachFromDisposeScope() as Parameter;
```
Comment (shaltielshmid) [inline src/TorchSharp/NN/Module.cs:256]: I may be wrong about this - If I understood the Parameter class correctly, the handle is moved from the Tensor to the Parameter on creation (and the Tensor's handle is set to 0, so we don't care if it's disposed), and therefore we don't need to detach the internal tensor, just the final parameter.

**Comment by shaltielshmid** [comment]
Also - that extra test you added for the bug, were there any changes done to fix it?

File: test/TorchSharpTest/NN.cs (hunk 1 above)
Comment (NiklasGustafsson) [reply test/TorchSharpTest/NN.cs:4722]: But that wasn't the actual issue. 

The issue was that a) it used IntPtr.Zero in the base constructor call, and then (once that was addressed) b) that the 'weight' and 'bias' properties weren't set, so the parameters fetched by the properties were the old ones. The ComponentName attribute fixed that.

**Comment by NiklasGustafsson** [comment]
> Also - that extra test you added for the bug, were there any changes done to fix it?

No, but I liked testing that indexing with an empty tensor filter works. :-)

File: src/TorchSharp/NN/Module.cs (hunk 2 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Module.cs:256]: Yeah... better safe than sorry.

Comment (shaltielshmid) [reply src/TorchSharp/NN/Module.cs:256]: Fair enough :)

File: test/TorchSharpTest/NN.cs (hunk 1 above)
Comment (shaltielshmid) [reply test/TorchSharpTest/NN.cs:4722]: Got it. So maybe just add a check to clarify that it isn't invalid?

```cs
Assert.False(pool.IsInvalid)
```

Comment (NiklasGustafsson) [reply test/TorchSharpTest/NN.cs:4722]: `Module.handle` is not public, so no can do. However, any call to `to()` will blow up if it's invalid, so the fact that the call does not should prove the same point.

Comment (shaltielshmid) [reply test/TorchSharpTest/NN.cs:4722]: Good point, I missed that part

**Review by shaltielshmid — APPROVED** [review]
